lib/rewrite/rewrite-set-facility.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2020 Balazs Scheidler <bazsi77@gmail.com>
+ * Copyright (c) 2019 Balabit
+ * Copyright (c) 2019 Kokan <kokaipeter@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "rewrite-set-facility.h"
+#include "template/templates.h"
+#include "syslog-names.h"
+#include "scratch-buffers.h"
+#include "parse-number.h"
+
+#include <stdlib.h>
+
+typedef struct _LogRewriteSetFacility LogRewriteSetFacility;
+
+struct _LogRewriteSetFacility
+{
+  LogRewrite super;
+  LogTemplate *facility;
+};
+
+static gint
+_convert_facility_as_number(const gchar *facility_text)
+{
+  gint64 facility = 0;
+
+  if (!parse_dec_number(facility_text, &facility))
+    return -1;
+
+  if (facility > 127)
+    return -1;
+
+  return FACILITY_CODE(facility);
+}
+
+static gint
+_convert_facility_as_text(const gchar *facility_text)
+{
+  return syslog_name_lookup_facility_by_name(facility_text);
+}
+
+static gint
+_convert_facility(const gchar *facility_text)
+{
+  gint facility = _convert_facility_as_number(facility_text);
+  if (facility >= 0)
+    return facility;
+
+  facility = _convert_facility_as_text(facility_text);
+  if (facility >= 0)
+    return facility;
+
+  return -1;
+}
+
+static void
+_set_msg_facility(LogMessage *msg, const guint32 facility)
+{
+  msg->pri = (msg->pri & ~LOG_FACMASK) | facility;
+}
+
+static void
+log_rewrite_set_facility_process(LogRewrite *s, LogMessage **pmsg, const LogPathOptions *path_options)
+{
+  ScratchBuffersMarker marker;
+  GString *result = scratch_buffers_alloc_and_mark(&marker);
+  LogRewriteSetFacility *self = (LogRewriteSetFacility *) s;
+
+  log_msg_make_writable(pmsg, path_options);
+
+  log_template_format(self->facility, *pmsg, NULL, LTZ_LOCAL, 0, NULL, result);
+
+  const gint facility = _convert_facility(result->str);
+  if (facility < 0)
+    {
+      msg_debug("Warning: invalid value passed to set-facility()",
```

Comment:
This should be an error message.

Reply:
This is the same as in set-severity () where I asked to make this debug.

This is a per message path and I would rather ignore this as an error than generate shitloads of errors in case a parser goes wrong or it an attacker generates incorrect value on purpose.

We should not trust the value too much.